### PR TITLE
Response abstraction

### LINKED ISSUE #23: Class abstraction for PayboxResponse
Try to remove magic calls on `PayboxResponse` and do abstraction like on request classes.

### PR DESCRIPTION
**Pull request by soullivaneuh**
Closes #23.

### CODE REVIEW

File: src/Paybox.php [hunk 1]
```diff
@@ -48,15 +49,49 @@ public function __construct(array $options = [], AbstractHttpClient $httpClient
     /**
      * @param RequestInterface $request
      *
-     * @return PayboxResponse
+     * @return DirectResponse
+     */
+    public function requestDirect(RequestInterface $request)
```
Comment (greg0ire) [inline src/Paybox.php]: This method name does not feel right, it looks like you're "requesting a direct", which is not very meaningful… maybe `sendDirectRequest` and `sendDirectPlusRequest`

File: src/HttpClient/AbstractHttpClient.php [hunk 2]
```diff
@@ -100,7 +101,7 @@
             throw new PayboxException($results['COMMENTAIRE'], $results['CODEREPONSE']);
         }
 
-        return new PayboxResponse($results);
+        return new $responseClass($results);
```
Comment (greg0ire) [inline src/HttpClient/AbstractHttpClient.php:108]: Maybe check `$responseClass` implements `ResponseInterface` (with [class_implements](http://php.net/manual/en/function.class-implements.php) ?)

File: src/Response/AbstractResponse.php [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        $this->code = intval($data['CODEREPONSE']);
+        $this->comment = $data['COMMENTAIRE'];
+        $this->site = $data['SITE'];
+        $this->rank = $data['RANG'];
+        $this->callNumber = intval($data['NUMAPPEL']);
+        $this->questionNumber = intval($data['NUMQUESTION']);
+        $this->transactionNumber = intval($data['NUMTRANS']);
+
+        if (array_key_exists('AUTORISATION', $data)) {
+            $this->authorization = '' === $data['AUTORISATION'] ? false : $data['AUTORISATION'];
```
Comment (greg0ire) [inline src/Response/AbstractResponse.php]: A call to `array_filter` could save you a few checks like this.

File: src/Response/AbstractResponse.php [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        }
+        if (array_key_exists('PAYS', $data)) {
+            $this->country = in_array($data['PAYS'], ['', '???'], true) ? false : $data['PAYS'];
```
Comment (greg0ire) [inline src/Response/AbstractResponse.php]: What is this `???` about ?

File: src/Response/DirectPlusResponse.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ */
+final class DirectPlusResponse extends AbstractResponse
+{
+    /**
+     * @var string
+     */
+    private $subscriberRef;
+
+    /**
+     * @var string|false False if empty.
```
Comment (greg0ire) [inline src/Response/DirectPlusResponse.php:16]: Doesn't `null` better represent the absence of value?

File: src/Response/AbstractResponse.php (hunk 4 above)
Comment (soullivaneuh) [inline src/Response/AbstractResponse.php]: From Paybox documentation (FR):

```
PAYS
Format : 3 caractères (code ISO3166 alphabétique)
Code pays du porteur de la carte. La valeur « ??? » sera retournée si le code pays est inconnu.
Exemple : FRA
```

File: src/Response/DirectPlusResponse.php (hunk 5 above)
Comment (soullivaneuh) [inline src/Response/DirectPlusResponse.php:16]: Well, I decided to follow same logic is this: https://github.com/nexylan/paybox-direct/pull/32/files#diff-e06a2f8820ff95ef72cc9e316326ad9dR46
- `null` when the property is not returned by Paybox
- `false` when it's return but invalid (empty or `???` for country)

In this case, the bearer is **always** returned but can sometime be empty (on subscriber delete if i'm right.).

Comment (greg0ire) [inline src/Response/DirectPlusResponse.php:16]: Oh ok then.

File: src/HttpClient/AbstractHttpClient.php (hunk 2 above)
Comment (soullivaneuh) [inline src/HttpClient/AbstractHttpClient.php:108]: Done.

File: src/Response/AbstractResponse.php (hunk 4 above)
Comment (greg0ire) [inline src/Response/AbstractResponse.php]: pre-natal facepalm

File: src/HttpClient/AbstractHttpClient.php [hunk 6]
```diff
@@ -76,6 +76,10 @@
      */
     public function call($type, array $parameters, $responseClass)
     {
+        if (!in_array(ResponseInterface::class, class_implements($responseClass))) {
+            throw new \InvalidArgumentException('The response class must implements '.ResponseInterface::class.'.');
```
Comment (greg0ire) [inline src/HttpClient/AbstractHttpClient.php]: "must implement"

**Pull request merged**